File: 2020/cpp/day_13.cpp
#include "day_13.h"

day_13::day_13() {
    auto input_strings = readFile("Inputs/day_13.txt");
    const auto start = std::chrono::high_resolution_clock::now();
    starting_time = fast_atol(input_strings[0].data());
    auto splitted = split(input_strings[1], ',');
    for (const auto sv : splitted) {
        if (sv[0] == 'x') {
            all_bus_routes.push_back(-1);
            continue;
        }
        const auto val = fast_atol(sv.data());
        meaningful_bus_routes.push_back(val);
        all_bus_routes.push_back(val);
    }
    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("Parsing input for day thirteen took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

void day_13::part_one() const {
    const auto start = std::chrono::high_resolution_clock::now();
    auto i = starting_time;
    while (true) {
        for (auto route : meaningful_bus_routes) {
            if (i % route == 0) {
                const auto wait_time = i - starting_time;
                auto end = std::chrono::high_resolution_clock::now();
                fmt::print("The answer for day thirteen part one is {}\n", route * wait_time);
                fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
                return;
            }
        }
        ++i;
    }
}

void day_13::part_two() const {
    const auto start = std::chrono::high_resolution_clock::now();
    std::vector<std::pair<uint64_t, uint64_t>> wait_times;
    int minute{-1};
    uint64_t multiplier{0};
    for (const auto& bus : all_bus_routes) {
        ++minute;
        if (bus == -1) {
            continue;
        }
        if (minute == 0) {
            multiplier = bus;
        } else {
            wait_times.emplace_back(bus, bus - (minute % bus));
        }
    }
    uint64_t timestamp = 0;

    for (auto [first, second] : wait_times) {
        while ((timestamp % first) != second) {
            timestamp += multiplier;
        }
        multiplier *= first;
    }
    auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day thirteen part one is {}\n", timestamp);
    fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

File: 2020/rust/src/day13.rs
use std::time::{Duration, Instant};

struct Timetable {
    depart_time: i64,
    buses: Vec<Option<i64>>,
}

fn parse_input(input: &str) -> (Timetable, Duration) {
    let start = Instant::now();
    let mut lines = input.lines();
    let depart_time = lines.next().and_then(|x| x.parse().ok()).unwrap();
    let valid_buses = lines
        .next()
        .unwrap()
        .split(',')
        .map(|x| x.parse().ok())
        .collect();
    let out = Timetable {
        depart_time,
        buses: valid_buses,
    };
    let end = Instant::now();
    (out, end - start)
}

fn solve_part1(input: &Timetable) -> (i64, Duration) {
    let start = Instant::now();

    let mut bus = 0;
    let mut actual_time = 0;

    'outer: for time in input.depart_time.. {
        for bus_id in input.buses.iter().flatten() {
            if time % bus_id == 0 {
                bus = *bus_id;
                actual_time = time;
                break 'outer;
            }
        }
    }

    let out = bus * (actual_time - input.depart_time);

    let end = Instant::now();
    (out, end - start)
}

fn solve_part2(input: &Timetable) -> (i64, Duration) {
    let start = Instant::now();
    let mut wait_times = Vec::with_capacity(input.buses.len());
    let mut minute: i64 = -1;
    let mut multiplier = 0;

    for bus in input.buses.iter() {
        minute += 1;
        if let Some(bus) = bus {
            if minute == 0 {
                multiplier = *bus;
            } else {
                wait_times.push((bus, bus - (minute % bus)));
            }
        }
    }

    let mut timestamp = 0;
    for (first, second) in wait_times {
        while (timestamp % first) != second {
            timestamp += multiplier;
        }
        multiplier *= first;
    }

    let end = Instant::now();
    (timestamp, end - start)
}

pub fn run(input: String) {
    let (parsed_input, time) = parse_input(&input);
    println!("Day 13, parsing input took {}ns", time.as_nanos());
    let (p1, time) = solve_part1(&parsed_input);
    println!("Day 13, part 1: {}. Took {}ns", p1, time.as_nanos());
    let (p2, time) = solve_part2(&parsed_input);
    println!("Day 13, part 2: {}. Took {}ns", p2, time.as_nanos());
}
